### ISSUE TITLE
async foundations

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/lang-team#33]
# Summary

To pursue foundational work related to Async I/O. Much of this work involves extending the language with new foundational concepts. 

The group also pursues work that is unrelated to the lang team, such as compiler team polish.

# Info

* Lead: @nikomatsakis, @tmandry 
* Lang-team liaison: @nikomatsakis 
* [Charter](https://github.com/rust-lang/wg-async-foundations#what-is-the-goal-of-this-working-group)
* Repository [rust-lang/wg-async-foundations](https://github.com/rust-lang/wg-async-foundations)
* Zulip stream `#wg-async-foundations`

# What is this issue?

This issue represents an **active project group**. It is meant to be used for
the group to post updates to the lang team (and others) in a lightweight
fashion. Please do not use the comments here for discussion, that should be kept
in the Zulip stream (discussion comments here will be marked as off-topic).

**Comment by nikomatsakis** [comment]
Update 2020-07-27:

* Still working towards the Stream RFC, should schedule the meeting about it.
* boats proposed `AsyncDrop`, withdrawn but may come up later.

**Comment by nikomatsakis** [comment]
Update 2020-08-03:

* Ongoing work on the `Stream` trait, meeting likely next week

**Comment by nikomatsakis** [comment]
* Stream trait meeting planned for this Wednesday.

**Comment by nikomatsakis** [comment]
Update 2020-08-31:

* Had the meeting, no major blockers identified.
* Planning to open the RFC soon, but will add a plan for migration.

**Comment by nikomatsakis** [comment]
Update 2020-09-14:

* Discussed how the `Stream` RFC is really more of a "libs with lang consultation" RFC than anything else.
* Looking towards a lint similar to `must_use` but about values held live across yields.

**Comment by nikomatsakis** [comment]
Update 2020-09-21:

* Still waiting on `Stream` RFC.

**Comment by joshtriplett** [comment]
Stream RFC: https://github.com/rust-lang/rfcs/pull/2996

**Comment by joshtriplett** [comment]
This shouldn't need further lang review until after that RFC has been merged and implemented.

**Comment by nikomatsakis** [comment]
Update 2020-10-07:

There is a [Draft RFC available](https://github.com/rust-lang/wg-async-foundations/pull/16/files) for a `must_not_await` lint.

**Comment by nikomatsakis** [comment]
Update 2021-04-06:

* Working on the [async vision doc]()https://rust-lang.github.io/wg-async-foundations/vision.html). Will be coordinating with lang team as we get a clearer idea of the roadmap and "shiny future" proposals.
* Another point of upcoming contact will be about the ["async guiding principles"](https://hackmd.io/haMJf784SAemBbGyO5Wi6Q?edit), which -- at least at present -- look like they may be an interesting "specialization" of more general Rust language guiding principles. **Good topic for a design meeting, perhaps.**

**Comment by nikomatsakis** [comment]
Update 2021-07-07:

* The async vision doc has reached the end of its brainstorming period.
* We've been iterating towards what we believe to be the "shiny future", and there's a lot to discuss.
* You can find an overview of the current status [in this diagram](http://smallcultfollowing.com/async-harmonic-synthesis.svg), which has links to more detailed design notes. 

Most likely the best thing would be to have a design meeting to talk it over.

**Comment by nikomatsakis** [comment]
Closing in favor of https://github.com/rust-lang/async-fundamentals-initiative/issues/5
